core/src/main/java/com/scalar/db/api/ScanBuilder.java

```diff
@@ -548,4 +873,538 @@ public Scan build() {
+     *
+     * @param condition a conditional expression for a {@code OrConditionSet}.
+     * @return a builder object
+     */
+    public BuildableOrConditionSet or(ConditionalExpression condition) {
+      conditions.add(condition);
+      return new BuildableOrConditionSet(conditions);
+    }
+  }
+
+  public static class BuildableAndConditionSet {
+
+    private final Set<ConditionalExpression> conditions;
+
+    private BuildableAndConditionSet(Set<ConditionalExpression> conditions) {
+      this.conditions = new HashSet<>();
+      this.conditions.addAll(conditions);
+    }
+
+    /**
+     * Adds a conditional expression for a {@code AndConditionSet}.
+     *
+     * @param condition a conditional expression for a {@code AndConditionSet}.
+     * @return a builder object
+     */
+    public BuildableAndConditionSet and(ConditionalExpression condition) {
+      conditions.add(condition);
+      return this;
+    }
+
+    /**
+     * Builds an and-wise condition set with the specified conditional expressions.
+     *
+     * @return an and-wise condition set
+     */
+    public AndConditionSet build() {
+      return new AndConditionSet(ImmutableSet.copyOf(conditions));
+    }
+  }
+
+  public static class BuildableOrConditionSet {
+
+    private final Set<ConditionalExpression> conditions;
+
+    private BuildableOrConditionSet(Set<ConditionalExpression> conditions) {
+      this.conditions = new HashSet<>();
+      this.conditions.addAll(conditions);
+    }
+
+    /**
+     * Adds a conditional expression for a {@code OrConditionSet}.
+     *
+     * @param condition a conditional expression for a {@code OrConditionSet}.
+     * @return a builder object
+     */
+    public BuildableOrConditionSet or(ConditionalExpression condition) {
+      conditions.add(condition);
+      return this;
+    }
+
+    /**
+     * Builds an or-wise condition set with the specified conditional expressions.
+     *
+     * @return an or-wise condition set
+     */
+    public OrConditionSet build() {
+      return new OrConditionSet(ImmutableSet.copyOf(conditions));
+    }
+  }
+
+  private static Scan buildScanAllWithConjunctions(
+      String namespaceName,
+      String tableName,
+      @Nullable ConditionalExpression condition,
+      Set<Set<ConditionalExpression>> conjunctions,
+      Set<Set<ConditionalExpression>> disjunctions,
+      List<String> projections,
+      List<Scan.Ordering> orderings,
+      int limit,
+      @Nullable com.scalar.db.api.Consistency consistency) {
+    ScanAll scan = new ScanAll();
+    scan.forNamespace(namespaceName).forTable(tableName).withLimit(limit);
+    orderings.forEach(scan::withOrdering);
+
+    if (condition != null) {
+      assert conjunctions.isEmpty() && disjunctions.isEmpty();
+      scan.withConjunctions(ImmutableSet.of(Scan.Conjunction.of(condition)));
+    } else if (conjunctions.isEmpty()) {
+      scan.withConjunctions(
+          Sets.cartesianProduct(new ArrayList<>(disjunctions)).stream()
```

Comment:
(Sorry for the delayed comment. I thought I put the following comment here, but it's gone...)

This seems to generate O x A conditions (O sets of A conditions), where O is the number of ORs and A is the number of ANDs. I'm not sure how these sets of conditions would be used in next PRs, but I guess ScalarDB fetches all entries and then filter the fetched records using these conditions set.

(This would be a silly question..., but) it's not possible to pushdown original conditions basically as-is to underlying RDBMS?

Reply:
Thank you for the question. It will be used in something like the following way. We don't filter results using the conditions (at least for JDBC storage) in the ScalarDB layer.

```java
  private String relationalConditionSqlString() {
   (snip)
    List<String> conjunctionList = new ArrayList<>();
    conjunctions.forEach(
        conjunction -> {
          List<String> conditions = new ArrayList<>();
          conjunction
              .getConditions()
              .forEach(
                  condition ->
                      conditions.add(condition.getColumn().getName() + convert(condition.getOperator())));
          conjunctionList.add(String.join(" AND ", conditions));
        });
    return " WHERE " + String.join(" OR ", conjunctionList);
  }
```

> (This would be a silly question..., but) it's not possible to pushdown original conditions basically as-is to underlying RDBMS?

Do you mean pushing down both DNF and CNF as is? It's technically possible, and it sometimes would perform better, but I think it trades off a development effort and maintainability for two internal representations.

Reply:
Thanks for the answer. It sounds good!

What I meant was...
converting the following Scan
```
    Scan.newBuilder()
        .namespace("my_ns")
        .table("my_tbl")
        .all()
        .where(ConditionBuilder.column("c1").isEqualToInt(10))
        .and(ConditionBuilder.column("c2").isEqualToBigInt(20))
        .and(
            ConditionSetBuilder.condition(
                ConditionBuilder.column("c3").isEqualToInt(30))
                .or(ConditionBuilder.column("c4").isEqualToInt(40))
                .build())
           :
        .build();
```
into the following SQL
```
SELECT ...
FROM my_ns.my_tbl
WHERE c1 = 10 AND c2 = 20 AND (c3 = 30 OR c4 = 40)
```
although I'm not sure it works well 😅

Reply:
I see. That's pushing down a CNF as is. Yeah, as I said, technically, it works if we make `Scan` always have either conjunctions or disjunctions. But my concern is that we probably need to do the same for the gRPC proto and take care of which one should be used in many places. Anyway, this is kind of an internal part; we can revisit this in the future. Any additional discussions are welcome. Thanks for the comment.